### DISCUSSION TITLE
How to use a context variable with contains() in an if: line?

### DISCUSSION THREAD

**Discussion by gitcos** [community/community#25252] [Actions]
<p>I’m trying to figure out how to use if: to make a job run based on whether a certain variable from the context contains a particular substring.  Unfortunately, the documentation for if: doesn’t show any examples of actually using expressions with variables in them, and the documentation for contains() doesn’t show examples of how it would be used in an if:, and every guess I make about how to combine these things seems to fail.</p>
<p><code>if: ${{ contains('string', 'x') }}</code> skips the job. Good.<br>
<code>if: ${{ contains('string', 'g') }}</code> runs the job. Good.<br>
<code>echo ${{ github.action }}</code> always echoes <code>run1</code> so I decided to use that as my test value…</p>
<p>I saw a post on a forum saying that you don’t actually need to use ${{ }} in if because it just knows it’s an expression and access context variables, so I tried both with and without curly braces:</p>
<p><code>if: contains(github.action, 'r')</code> skips the job! So that doesn’t work.</p>
<p><code>if: contains(${{ github.action }}, 'r')</code> runs the job. Okay, seems promising…<br>
… but <code>if: </code>contains(${{ github.action }}, 'ZZZ<code>)</code> also runs the job! So this is just returning true no matter what string I give it?</p>
<p>The opposite is true when I put curly braces around the whole expression,<br>
<code>if: ${{ contains(github.action, 'r') }}</code> and <code>if: ${{ contains(github.action, 'ZZZ') }}</code> both skip the job, so that one is always false.</p>
<p>So what’s going on?  How can I actually use <code>if:</code> to make a job run only if a variable contains a particular substring?</p>
<p>[ Sorry if this is a repeat. The search feature on this forum doesn’t work. For example, when I search for ‘contains’ it shows me lots of posts about Docker containers because it thinks ‘contains’ and ‘container’ are the same thing, or something. I don’t know, I just know that I failed to get relevant results for most of my searches. ]</p>

**Comment by kingthorin** [comment]
<p><code>if: contains('["kingthorin","lpoulter"]', github.actor)</code></p>
<p>That’s from another post of mine, hopefully it leads you to a solution. Basically it checks if the github.actor is one of those in the array or string literals.</p>

**Comment by gitcos** [comment]
<p>I’m specifically trying to check a substring of a string, not look for elements in an array.</p>
<p>However, if we leave that difference aside, I think you’re telling me that there is definitely no need for ${{ }} anywhere on the if: line to do this. If that’s true, even though it doesn’t give me the full answer, it does limit what I should be trying, so it would be very helpful to have that confirmed. Do you know, confidently, that that is true?</p>
<p>Secondly, the documentation for contains() on <a href="http://github.com" rel="noopener nofollow ugc">github.com</a> it says:</p>
<blockquote>
<p>contains</p>
<p><code>contains( search, item )</code></p>
<p>Returns <code>true</code> if <code>search</code> contains <code>item</code> . If <code>search</code> is an array, this function returns <code>true</code> if the <code>item</code> is an element in the array. If <code>search</code> is a string, this function returns <code>true</code> if the <code>item</code> is a substring of <code>search</code> . This function is not case sensitive. Casts values to a string.</p>
</blockquote>
<p>Your example looks like the opposite of what that documentation says.  In your example, it’s looking for whether <code>search</code> <em>is contained in</em> <code>item</code>, not whether search <strong>contains</strong> item.  Are you saying the documentation is wrong, and got it backwards?</p>
<p>(Sorry, I tried to link to the documentation I’m quoting from, but this forum apparently doesn’t allow links so it forced me to take that out)</p>

**Comment by gitcos** [comment]
<aside class="quote no-group" data-username="gitcos" data-post="1" data-topic="235210">
<div class="title">
<div class="quote-controls"></div>
 gitcos:</div>
<blockquote>
<p><code>if: contains(${{ github.action }}, 'r')</code> runs the job. Okay, seems promising…<br>
… but <code>if: </code> contains(${{ github.action }}, 'ZZZ <code>)</code> also runs the job! So this is just returning true no matter what string I give it?</p>
</blockquote>
</aside>
<p>Unfortunately this forum does not let me edit my post, so to fix a small formatting error, here’s what that paragraph should say:</p>
<aside class="quote no-group">
<blockquote>
<p><code>if: contains(${{ github.action }}, 'r')</code> runs the job. Okay, seems promising…<br>
… but <code>if: contains(${{ github.action }}, 'ZZZ ')</code> also runs the job! So this is just returning true no matter what string I give it?</p>
</blockquote>
</aside>

**Comment by kingthorin** [comment]
<p>Here’s an example for you:</p><aside class="onebox githubactions" data-onebox-src="https://github.com/kingthorin/testwf/actions/runs/1970537519">
  <header class="source">

      <a href="https://github.com/kingthorin/testwf/actions/runs/1970537519" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>

  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Commit">
      <svg width="24" height="24" class="github-icon github-icon-success" viewbox="0 0 24 24" aria-hidden="true"><path d="M1 12C1 5.925 5.925 1 12 1s11 4.925 11 11-4.925 11-11 11S1 18.075 1 12zm16.28-2.72a.75.75 0 00-1.06-1.06l-5.97 5.97-2.47-2.47a.75.75 0 00-1.06 1.06l3 3a.75.75 0 001.06 0l6.5-6.5z"></path></svg>



  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/kingthorin/testwf/actions/runs/1970537519" target="_blank" rel="noopener nofollow ugc">Update test.yml</a>
    </h4>

      <div class="github-info">
        Test <span class="github-run-number">#6</span>
      </div>
  </div>
</div>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>Confirming this syntax: <code>if: contains('jane kingthorin fred', github.actor)</code></p>

**Comment by gitcos** [comment]
<p>Ahh, okay, I see what you’re doing - instead of trying to check whether a substring is found inside the value of a variable, you’re doing the opposite: checking whether the value of a variable is found inside a longer strong.</p>
<p>Does this syntax work for you if you try to do what I want to do?  For example, if you try:</p>
<blockquote>
<p>if: contains(github.actor, ‘ed’)</p>
</blockquote>
<p>Will that run your job?</p>
<p>And then, if you change it to:</p>
<blockquote>
<p>if: contains(github.actor, ‘X’)</p>
</blockquote>
<p>Will that skip your job and not run it?</p>
<p>If those two examples work like that for you, then that shows the syntax works for you, and means the problem may be something specific to my environment, or server, or workflow.  That would be very helpful to know!</p>

**Comment by kingthorin** [comment]
<p>Double checked: <code>if: contains('jane fred', github.actor)</code> you can see it didn’t actually do anything: <a href="https://github.com/kingthorin/testwf/runs/5515857258?check_suite_focus=true" class="inline-onebox" rel="noopener nofollow ugc">Test · kingthorin/testwf@acbae8b · GitHub</a> (the middle step is skipped since the if isn’t fulfilled).</p>

**Comment by kingthorin** [comment]
<p>I can test those for you in a few hours. Will post an update then :wink:</p>

**Comment by gitcos** [comment]
<aside class="quote no-group" data-username="kingthorin" data-post="8" data-topic="235210" data-full="true">
<div class="title">
<div class="quote-controls"></div>
 kingthorin:</div>
<blockquote>
<p>Double checked: <code>if: contains('jane fred', github.actor)</code> you can see it didn’t actually do anything: <a href="https://github.com/kingthorin/testwf/runs/5515857258?check_suite_focus=true" rel="noopener nofollow ugc">Test · kingthorin/testwf@acbae8b · GitHub</a> (the middle step is skipped since the if isn’t fulfilled).</p>
</blockquote>
</aside>
<p>I don’t understand why.  github.actor’s value is “fred”, and “fred” <strong>is</strong> contained in the string ‘jane fred’, so that if should resolve to true and not skip the job.  Can you explain why it’s false?</p>

**Comment by kingthorin** [comment]
<p>No no the input isn’t really used. GitHub.actor is my username because I started the workflow.</p>

**Comment by gitcos** [comment]
<p>What do you mean “the input isn’t really used”, and what is this comment responding to?  I’m very confused by this comment :slight_smile:</p>

**Comment by kingthorin** [comment]
<p>The input is only used to define who is being welcomed (echoed) by the workflow. I’ll modify it and do your tests.</p>

**Comment by kingthorin** [comment, marked as answer]
<details>
<summary>String with kingthorin contains github.actor (kingthorin)</summary>
<pre data-code-wrap="yml"><code class="lang-nohighlight">name: Test

on:
  workflow_dispatch:

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
    - name: Say Hi
      if: contains('jane kingthorin fred', github.actor)
      run: echo fred
</code></pre>
</details>
<p><a href="https://github.com/kingthorin/testwf/actions/runs/1971324173" rel="noopener nofollow ugc">Result</a> &gt; Workflow runs.</p>
<details>
<summary>String without kingthorin contains github.actor (kingthorin)</summary>
<pre data-code-wrap="yml"><code class="lang-nohighlight">name: Test

on:
  workflow_dispatch:

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
    - name: Say Hi
      if: contains('jane fred', github.actor)
      run: echo fred
</code></pre>
</details>
<p><a href="https://github.com/kingthorin/testwf/actions/runs/1971331063" rel="noopener nofollow ugc">Result</a> &gt; Workflow runs skipping step, if is not fulfilled.</p>
<details>
<summary>github.actor contains 'king'</summary>
<pre data-code-wrap="yml"><code class="lang-nohighlight">
name: Test

on:
  workflow_dispatch:

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
    - name: Say Hi
      if: contains(github.actor, 'king')
      run: echo fred
</code></pre>
</details>
<p><a href="https://github.com/kingthorin/testwf/actions/runs/1971341232" rel="noopener nofollow ugc">Result</a> &gt; Workflow runs.</p>
<details>
<summary>github.actor contains 'jane'</summary>
<pre data-code-wrap="yml"><code class="lang-nohighlight">name: Test

on:
  workflow_dispatch:

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
    - name: Say Hi
      if: contains(github.actor, 'jane')
      run: echo fred
</code></pre>
</details>
<p><a href="https://github.com/kingthorin/testwf/actions/runs/1971346024" rel="noopener nofollow ugc">Result</a> &gt; Workflow run skipping step, if is not fulfilled.</p>

**Reply by kristoflovasz** [reply]
Hi!
I'm using contains with two lists ` contains(fromJSON('["failure", "cancelled"]'), needs.*.result) `
Can you explain how does this work under the hood, please?
Thank you very much!

**Reply by jsoref** [reply]
the `*` stuff is an [Object filter](https://docs.github.com/en/actions/learn-github-actions/expressions#object-filters)

**Comment by kingthorin** [comment]
<p><a class="mention" href="/gitcos">@gitcos</a> did that help you out? Is your issue resolved?</p>

**Comment by gitcos** [comment]
<p>Thanks!  That makes it clear that this syntax is supposed to work, and that ${{ }} should not be used.</p>
<p>I still don’t get what you were talking about with “the input isn’t really used” above - I don’t know what “input” you were referring to, I don’t know what you thought I thought that input is used for, etc.  I really don’t get anything at all whatsoever about that sentence or what it’s talking about, still.</p>
<p>However, this latest set of examples is pretty clear and comprehensive and shows that this syntax works.  Thank you for that!</p>

**Comment by kingthorin** [comment]
<p>Let me see if I can clarify.</p>
<p>You had said:</p>
<aside class="quote no-group" data-username="gitcos" data-post="10" data-topic="235210">
<div class="title">
<div class="quote-controls"></div>
 gitcos:</div>
<blockquote>
<p>I don’t understand why. github.actor’s value is “fred”, and “fred” <strong>is</strong> contained in the string ‘jane fred’, so that if should resolve to true and not skip the job. Can you explain why it’s false?</p>
</blockquote>
</aside>
<p>To which I had responded:</p>
<aside class="quote no-group" data-username="kingthorin">
<div class="title">
<div class="quote-controls"></div>
 kingthorin:</div>
<blockquote>
<p>No no the input isn’t really used. GitHub.actor is my username because I started the workflow.</p>
</blockquote>
</aside>
<p>At that point there had been an input to the workflow which had the default value “Fred” but that had nothing to do with the value of GitHub.actor. The input defaulted to Fred had just been used to identify whose name was going to be echoed in the workflow output. (It wasn’t relevant to your question and obviously threw you off so I removed it in later examples.)</p>
<p>Anyway if that latest set of examples answers your question please select it as the solution to this thread/question :wink:</p>

**Comment by gitcos** [comment]
<p>Oh, what you’re saying is “no, github.actor’s value is not ‘fred’ in this example”</p>
<p>That answers what I asked in that comment.</p>
<p>I just didn’t get what you meant by an input or how it related to what I asked, but I guess it isn’t relevant.  Thanks!</p>

**Comment by hvithval** [comment]
I know this post is a bit old, but I have just crossed paths with it and the fact that it made clear you don't need to use the **${{}}** was extremely helpful.
In my case, I needed it to check whether the word QA and DEVELOPMENT were in the output from a previous step and it worked (but still took me long enought to get to this post and try without the ${{}}}.
The idea was to set a condition at a job level so it would only run if QA or DEVELOPMENT were in the OUTPUT of a previous step:

```
jobs:
  prepare-matrix:
    runs-on: terraform
    outputs:
      matrix: ${{ steps.set-matrix.outputs.JSON_RESPONSE}}
    steps:
    - name: "Checkout source code"
      uses: actions/checkout@v3
      with:
        fetch-depth: 0
    - name: set-matrix
      id: set-matrix
      run: |
        ls terraform/ | grep -v modules | jq -R '[.]' | jq -s -c 'add' > envs.json
        cat envs.json
        echo 'JSON_RESPONSE<<EOF' >> $GITHUB_OUTPUT
        cat envs.json >> $GITHUB_OUTPUT
        echo 'EOF' >> $GITHUB_OUTPUT

  terraform-apply-qa:
    if: contains(fromJson(needs.prepare-matrix.outputs.matrix), 'qa')
    runs-on: terraform
    needs: prepare-matrix
    environment: qa
    #if: contains( ${{ matrix.environment }}, 'qa')
    steps:
      - name: "Checkout source code"
        uses: actions/checkout@v3
        with:
          fetch-depth: 0
```

So as you can see in the `terrafrom-apply-qa` job I'm using the following condition and it worked, confirming **${{}}** are not needed.
```
if: contains(fromJson(needs.prepare-matrix.outputs.matrix), 'qa')
```

**Reply by jsoref** [reply]
I opened https://github.com/orgs/community/discussions/60448 about:
```
    #if: contains( ${{ matrix.environment }}, 'qa')
```
